packages/story-editor/src/app/taxonomy/taxonomyProvider.js

```diff
@@ -90,40 +91,58 @@ function TaxonomyProvider(props) {
         cacheFromEmbeddedTerms(terms),
         (slug) => taxonomiesBySlug[slug]?.restBase
       );
-      const initialSelectedSlugs = mapObjectVals(initialCache, (val) =>
-        Object.keys(val)
+      const initialSelectedTerms = mapObjectVals(initialCache, (val) =>
+        Object.values(val).map((term) => term.id)
       );
 
-      hasHydrationRunOnce.current = true;
       setTermCache(initialCache);
-      setSelectedSlugs(initialSelectedSlugs);
+      clearHistory();
+      updateStory({
+        properties: {
+          terms: initialSelectedTerms,
+        },
+      });
+      hasHydrationRunOnce.current = true;
     }
-  }, [terms, isStoryLoaded, taxonomies, setSelectedSlugs, setTermCache]);
+  }, [
+    terms,
+    isStoryLoaded,
+    taxonomies,
+    setTermCache,
+    clearHistory,
+    updateStory,
+  ]);
 
-  // With the freeform taxonomy input, we can have terms selected
-  // that may be in the process of being created or retrieved from
-  // the backend. Because of this, we sync up our local selected slugs
-  // with whatever cached terms are available at any given moment.
-  useEffect(() => {
-    if (!hasHydrationRunOnce.current) {
-      return;
-    }
+  const setTerms = useCallback(
+    (taxonomy, termIds = []) => {
+      updateStory({
+        properties: (story) => {
+          const newTerms =
+            typeof termIds === 'function'
+              ? termIds(story.terms[taxonomy.restBase])
+              : termIds;
 
-    const termEntries = Object.entries(selectedSlugs).map(
-      ([taxonomyRestBase, termSlugs = []]) => [
-        taxonomyRestBase,
-        termSlugs
-          .map((termSlug) => termCache[taxonomyRestBase]?.[termSlug]?.id)
-          .filter((id) => typeof id === 'number'),
-      ]
-    );
-    const updatedTerms = objectFromEntries(termEntries);
-    updateStory({
-      properties: {
-        terms: updatedTerms,
-      },
-    });
-  }, [updateStory, selectedSlugs, termCache]);
+          return {
+            ...story,
+            terms: {
+              ...story.terms,
+              [taxonomy.restBase]: newTerms,
+            },
+          };
+        },
+      });
+    },
+    [updateStory]
+  );
+
+  const addTermToSelection = useCallback(
+    (taxonomy, term) => {
+      setTerms(taxonomy, (ids) =>
```

Comment:
```suggestion
      setTerms(taxonomy, (ids = []) =>
```

If you don't have any terms yet this will crash the app without a default prop set as an array.